cloud_functions/climateiq_merge_output_cf/main.py

```diff
@@ -1,7 +1,98 @@
+# different chunks.
+@functions_framework.http
+def merge_scenario_predictions(request: flask.Request) -> tuple[str, int]:
+    try:
+        run_id, prediction_type, model_id, study_area_name = _get_args(
+            request, ("run_id", "prediction_type", "model_id", "study_area_name")
+        )
+    except ValueError as error:
+        return f"Bad request: {error}", 400
+
+    storage_client = storage.Client()
+    bucket = storage_client.bucket(BUCKET_NAME)
+
+    blobs = storage_client.list_blobs(
+        BUCKET_NAME, f"{run_id}/{prediction_type}/{model_id}/{study_area_name}"
+    )
+    chunk_ids, scenario_ids = _get_chunk_and_scenario_ids(blobs)
+    for chunk_id in chunk_ids:
+        output_file_name = (
+            f"{OUTPUT_DIR}/{run_id}/{study_area_name}/{prediction_type}/{chunk_id}.csv"
+        )
+        blob_to_write = bucket.blob(output_file_name)
+        with blob_to_write.open("w") as fd:
+            writer = csv.DictWriter(fd, fieldnames=["h3_index"] + scenario_ids)
+            writer.writeheader()
+            predictions_by_h3_index: dict[str, dict] = collections.defaultdict(dict)
+            for scenario_id in scenario_ids:
+                object_name = (
+                    f"{run_id}/{prediction_type}/{model_id}/"
+                    f"{study_area_name}/{scenario_id}/{chunk_id}.csv"
+                )
+                for row in _get_file_content(bucket, object_name):
+                    predictions_by_h3_index[row["h3_index"]][scenario_id] = row[
+                        "prediction"
+                    ]
+            for h3_index, predictions in predictions_by_h3_index.items():
+                predictions["h3_index"] = h3_index
+                # Output CSV will have the headers: h3_index, scenario_0, scenario_1...
+                writer.writerow(predictions)
+    return "Success", 200
+
+
+def _get_args(
+    request: flask.Request, arg_names: collections.abc.Iterable[str]
+) -> list[str]:
+    args = []
+    for arg_name in arg_names:
+        arg_value = request.args.get(arg_name)
+        if not arg_value:
+            raise ValueError(f"Missing arg {arg_name}")
+        args.append(arg_value)
+    return args
+
+
+def _get_chunk_and_scenario_ids(
+    blobs: list[storage.Blob],
+) -> tuple[list[str], list[str]]:
+    chunk_ids = set()
+    scenario_ids = set()
+    for blob in blobs:
+        match = re.match(CHUNK_FILE_NAME_PATTERN, blob.name)
+        # Ignore blobs that don't match the pattern.
+        if not match:
+            continue
+        chunk_ids.add(match.group("chunk_id"))
+        scenario_ids.add(match.group("scenario_id"))
+    return sorted(list(chunk_ids), key=str), sorted(list(scenario_ids), key=str)
+
+
+def _get_file_content(bucket: storage.Bucket, object_name: str) -> list[dict]:
+    blob = bucket.blob(object_name)
+    # If the specific blob doesn't exist (i.e., no prediction for given scenario_id and
```

Comment:
in this case, I think we should actually throw and error and not output any merged CSVs for any chunks.

for all scenarios, we should have valid predictions for all chunks. if not, something went wrong so we shouldnt ship any new data

@waltaskew to confirm

Reply:
I definitely agree we shouldn't ship data if some chunks are missing. We don't want to export data with holes in it! 

I think it's generally better to be brittle and fail loudly if data looks funny rather than try to recover ship the funny looking data.

Reply:
Changed to fail if a file isn't found. However, transactionally deleting all the files that have been written so far would be complex.

Reply:
Could we add a clean up step before triggering the failure that deletes any file paths with no scenario id (indicating a merged output)?

Reply:
I'd be less concerned about deleting files. As long as we know not to ship them, that's fine. The successful files could also be useful for debugging.
My hope would be that re-running once the issue is resolved would overwrite everything. Or folks could delete them manually. Either way, some manual intervention is going to be required to fix whatever bug caused the issue.